crates/relayer/src/chain/cosmos.rs

```diff
@@ -1933,31 +1933,118 @@ impl ChainEndpoint for CosmosSdkChain {
                     self.grpc_addr.clone(),
                 ),
             )
+            .map(|client| {
+                client.max_decoding_message_size(
+                    self.config().max_grpc_decoding_size.get_bytes() as usize
+                )
+            })
             .map_err(Error::grpc_transport)?;
 
-        client = client
-            .max_decoding_message_size(self.config().max_grpc_decoding_size.get_bytes() as usize);
+        if request.pagination.is_enabled() {
+            let mut results = Vec::new();
+            let mut page_key = Vec::new();
 
-        let request = tonic::Request::new(request.into());
+            let mut limit = request.pagination.get_limit();
 
-        let response = self
-            .block_on(client.packet_commitments(request))
-            .map_err(|e| Error::grpc_status(e, "query_packet_commitments".to_owned()))?
-            .into_inner();
+            loop {
+                crate::time!(
+                    "query_packet_commitments_loop_iteration",
+                    {
+                        "src_chain": self.config().id.to_string(),
+                    }
+                );
+                let mut raw_request =
+                    ibc_proto::ibc::core::channel::v1::QueryPacketCommitmentsRequest::from(
+                        request.clone(),
+                    );
 
-        let mut commitment_sequences: Vec<Sequence> = response
-            .commitments
-            .into_iter()
-            .map(|v| v.sequence.into())
-            .collect();
-        commitment_sequences.sort_unstable();
+                if let Some(pagination) = raw_request.pagination.as_mut() {
+                    pagination.key = page_key;
+                }
 
-        let height = response
-            .height
-            .and_then(|raw_height| raw_height.try_into().ok())
-            .ok_or_else(|| Error::grpc_response_param("height".to_string()))?;
+                let mut tonic_request = tonic::Request::new(raw_request);
+                tonic_request.set_timeout(Duration::from_secs(10));
```

Comment:
Just a note for us that we should eventually make this either:
a) Configurable (eg. `grpc_timeout` setting)
b) Inferred from the pagination, ie. use larger timeout for `Paginate::All` and smaller one otherwise